devito/passes/clusters/buffering.py

```diff
@@ -250,18 +257,30 @@ def callback(self, clusters, prefix, cache=None):
         if self.options['buf-fuse-tasks']:
             return processed
         for b in buffers:
-            if not (b.is_writeonly or b.is_readonly):
+            if b.is_writeonly:
+                # `b` might be written by multiple, potentially mutually-exclusive,
+                # equations. For example, two equations that have or will have
+                # complementary guards, hence only one will be execute. In such a
+                # case, we can split all equations over separate IterationSpaces
+                key0 = lambda: Stamp()
+            elif b.is_readonly:
+                # `b` is read multiple times -- this could just be the case of
+                # coupled equations, so we more cautiously perform a
+                # "buffer-wise" splitting of the IterationSpaces (i.e., only
+                # relevant if there are at least two read-only buffers)
+                stamp = Stamp()
+                key0 = lambda: stamp
```

Comment:
are lines 271-272 not doing the same thing with line 265?

Reply:
No 272 reuses the same stamp everytime while 265 creates a new one every time ke0 is called

Reply:
indeed

Reply:
Thanks for this. I feel I never got the Stamps right.

Reply:
you don't need to understand what Stamp achieves to pick up the difference here though

it's basically as Mathias said

the underlying idea is actually very straightforward. There are a few caveats but in essence: ** different stamp => no fusion possible **  (the converse is generally false btw, and I'd say also of couse)